### ISSUE TITLE
Long lag spikes on some projects

### ISSUE THREAD

**Issue by Xcalibur10** [godotengine/godot#78754]
### Godot version

4.0.3.stable(.mono)

### System information

Linux Mint / Windows 10, Vulkan, R5 3600, 2x8 GB RAM, RX 570 4 GB

### Issue description

Having a really frustrating issue. I'm new to Godot and considering this engine for my project. As I usually do, I download and try out demo projects. 

ON WINDOWS 10:
There is an Overgrowth Subway one. Opened it, ran it and it has insane lag spikes. When it runs well, it produces 60+ fps. But every second, I have around a second long lag spike.

![SubwayGodotLagSpike](https://github.com/godotengine/godot/assets/4464351/ae61b295-506b-4bf3-a73b-f97740e9aa26)

What I realized is, when it runs properly, it does the same with the editor. So it's going back and forth. Another project I tried had some similar issue, but came much later. In that case, I went to fullscreen, messed with resolution scaling (the project had a menu for that), and it took minutes to get to that issue.
I checked out the profiler, the spikes are there, Process Time is over 1 seconds when this happens. Important to note, CPU usage is pretty low, just a few percent, but GPU usage spikes to 100% when this happens. Also when the lag spike happens it affects my whole computer.
Running these projects standalone feels fine. The one that has the constant spikes runs perfectly if the editor isn't running. It affects the editor, stopping the project from running fixes the editor performance. 

Abandoned Spaceship project: runs well for a while, often over 100 fps, sometimes even around 200 depends on what's on screen.
![GodotLagSpike](https://github.com/godotengine/godot/assets/4464351/86d59844-1dca-41a5-82ec-673a8212c43c)

ON LINUX MINT:
Only tried the Overgrowth Subway scene. Sadly, framerate is low, but Godot reports 70+ fps while it feels like 20-30 often, except when the reported fps is around 140-160, jumps up when looking at walls.
No stuttering at all, the performance is poor, but very consistent. Same hardware.

### Steps to reproduce

-Running projects from editor that are somewhat heavier visually.
-Sometimes it takes time to kick in, then it never stops the lag spikes.

On Linux, it didn't happen so far.

### Minimal reproduction project

N/A

**Comment by Zireael07** [comment]
If you have the spikes only with the editor is on, can you try disabling remote scene tree in editor settings and re-checking?

**Comment by Xcalibur10** [comment]
> If you have the spikes only with the editor is on, can you try disabling remote scene tree in editor settings and re-checking?

Tried it, same result.

**Comment by Xcalibur10** [comment]
> Please add links to the projects that you've been trying. Also note that custom projects may lack some optimizations, which are not necessarily an issue (or not a critical issue at least) with the engine itself.

Here are the links:
This one has crazy spikes all the time. GPU goes to 100% consistently, temperature are perfectly fine, around 50-60°C and the spikes are there from the beginning. Framerates are normally 60+ fps.
[Overgrowth Subway](https://godotengine.org/asset-library/asset/1852)
This project runs smoothly, sometimes for long, long minutes. And not even sure if the lag spikes can be avoided or not, messing with in-game settings can cause it?
[Abandoned Spaceship](https://godotengine.org/asset-library/asset/1733)

Btw, changing visual settings in case of the first demo increases FPS if I reduce effects but the lag spikes won't change at all.

**Comment by bitsawer** [comment]
I also get these lag spikes on some demos, but only when running the project through editor. Closing editor and running the project directly on command line makes the lagging go away, so I think we might just be running out of GPU memory etc. as the editor and running demo are competing for resources? Haven't really measured how large footprint these projects actually have during runtime, so it could be something else. I have only 4 GB of memory on my GPU too, might be easy to run out these days.

I also have problems with this demo: https://github.com/Calinou/godot4-demo-desert-light, but only when running it in a large resolution window (like 1920x1200) and enabling TAA inside the app settings. The lagging starts when that is enabled and doesn't stop even if you disable TAA. Project restart is needed.

My specs:
Godot v4.1.rc (1f9e540f1) - Windows 10.0.19045 - Vulkan (Forward+) - dedicated NVIDIA GeForce GTX 970 (NVIDIA; 31.0.15.1694) - Intel(R) Core(TM) i7-4790K CPU @ 4.00GHz (8 Threads)

**Comment by Xcalibur10** [comment]
Hmmmm... yeah, Desert Light is quiet heavy even with basically not having textures. Shadows, lighting, volumetric effects are doing the job.

Does the very same thing. VRAM usage is high. Not sure if it's high enough to start doing this. But it's dancing on the edge. But what's really crazy is how it freezes my whole computer when those spikes happen. So maybe VRAM issue or the lack of it. Not sure if it's normal that it uses this much. And even if I close the scene in the editor, it doesn't help.

Even with lowering the settings to the very bottom while the demo is running. Weirdly, even if I forced the program to run at lowest setting from the beginning by changing the script that controls the settings, it still ran like this (only set the preset to Low).

![Godot 4 Desert Light in editor](https://github.com/godotengine/godot/assets/4464351/4f529d98-d6b4-430c-8cd7-72c3710d77ee)
I was about to post this, but then I was thinking about something. What about just run the project from the editor and quickly minimize the editor window. Guess what. It runs fine.

![Godot Desert Light Standalone (not debug) Build](https://github.com/godotengine/godot/assets/4464351/16a302ee-f168-4eb0-98d0-2a9aa8cb4e76)
Standalone build I get 30+ FPS, but there is no lag spike at all. But if I tab out to Godot editor then go back to the standalone program, it starts the lagging however, just clicking on something in the window fixes it.

So the Editor destroys performance. Constant lag spikes on heavier projects. Even with no profiling, local scene tree...

**Comment by Calinou** [comment]
> So the Editor destroys performance. Constant lag spikes on heavier projects. Even with no profiling, local scene tree...

When the editor is in the background (even minimized), VRAM still has to be allocated for both the editor and running project. On GPUs with low amounts of VRAM (by 2023 standards), this can be an issue.

**Comment by Calinou** [comment]
> I remember asking for auto-switch to script tab when running project for similar reasons, but can't find the issue number atm...

Switching tabs isn't enough to free all VRAM – you also need to close the currently open scenes in the editor (and reopen them once you're done running the project). Unfortunately, this can be a significant hindrance to workflow and prevents you from using live scene editing.

**Comment by Xcalibur10** [comment]
> > So the Editor destroys performance. Constant lag spikes on heavier projects. Even with no profiling, local scene tree...
> 
> When the editor is in the background (even minimized), VRAM still has to be allocated for both the editor and running project. On GPUs with low amounts of VRAM (by 2023 standards[1](#user-content-fn-1-7e95dacc3262ccfb34a760a7477beb73)), this can be an issue. To avoid this, close the scene in the 3D editor after running the project and switch to the Script tab after the project starts.
> 
> ## Footnotes
> 1. A GPU with only 4 GB of VRAM makes it difficult to get by in modern games/engines in 2023, even in 1080p. This is doubly true when two instances are running at the same time. [↩](#user-content-fnref-1-7e95dacc3262ccfb34a760a7477beb73)

Yeah, but even with the overall high VRAM usage, minimizing the editor fixes the issue. So it's not (just) VRAM issue. And it doesn't matter if the scene is open or not in the editor. I can start running the project, close the scene, if the editor window is still normal or maximized, it doesn't matter it has lag spikes.

On Linux, this issue doesn't exist for me. It just has an overall poor performance. Don't know if it's driver related, however, I played some games on it that has pretty high system requirements, even with Proton they ran like on Windows or better most of the time.

**Comment by Calinou** [comment]
> On Linux, this issue doesn't exist for me. It just has an overall poor performance. 

Windowed apps on Linux have to go through compositing, which slows down overall rendering performance. Try running the editor in fullscreen or better, disable compositing if your window manager allows it (e.g. <kbd>Alt + Shift + F12</kbd> with KWin).

**Comment by Xcalibur10** [comment]
> > On Linux, this issue doesn't exist for me. It just has an overall poor performance.
> 
> Windowed apps on Linux have to go through compositing, which slows down overall rendering performance. Try running the editor in fullscreen or better, disable compositing if your window manager allows it (e.g. Alt + Shift + F12 with KWin).

Hmmm... Mint is using Cinnamon Desktop Environment, don't know how to disable it. To be fair, not even sure if it makes difference in this case. I tried Godot 3.5 for a test, tried the [TPS Demo](https://github.com/godotengine/tps-demo) on it, and editor, 60 fps capped, gameplay with a little stuttering at the beginning, it works perfectly after a while, maybe it's shader caching or something else, but it runs smooth then at highest settings. I need to try this on Windows too.

**Comment by quinnyo** [comment]
From your screenshot (of Task Manager) it looks like Windows is allocating some system memory for use by the GPU (in addition to its actual VRAM), with some of it in use ("Shared GPU Memory"). It may not be getting used by the Godot project/editor, I can't say -- but if it is it will probably be pretty detrimental to performance.
